embassy-stm32/src/spi/v3.rs

```diff
@@ -161,9 +175,167 @@ impl<'d, T: Instance> Spi<'d, T> {
+        let f = self.txdma.write(request, write, dst);
+
+        unsafe {
+            T::regs().cfg1().modify(|reg| {
+                reg.set_txdmaen(true);
+            });
+            T::regs().cr1().modify(|w| {
+                w.set_spe(true);
+            });
+            T::regs().cr1().modify(|w| {
+                w.set_cstart(true);
+            });
+        }
+
+        f.await;
+        unsafe {
+            T::regs().cfg1().modify(|reg| {
+                reg.set_rxdmaen(false);
+                reg.set_txdmaen(false);
+            });
+            T::regs().cr1().modify(|w| {
+                w.set_spe(false);
+            });
+        }
+
+        Ok(())
+    }
+
+    #[allow(unused)]
+    async fn read_dma_u8(&mut self, read: &mut [u8]) -> Result<(), Error>
+    where
+        Tx: TxDmaChannel<T>,
+        Rx: RxDmaChannel<T>,
+    {
+        Self::set_word_size(WordSize::EightBit);
+        unsafe {
+            T::regs().cr1().modify(|w| {
+                w.set_spe(false);
+            });
+            T::regs().cfg1().modify(|reg| {
+                reg.set_rxdmaen(true);
+            });
+        }
+
+        let clock_byte_count = read.len();
+
+        let rx_request = self.rxdma.request();
+        let rx_src = T::regs().rxdr().ptr() as *mut u8;
+        let rx_f = self.rxdma.read(rx_request, rx_src, read);
+
+        let tx_request = self.txdma.request();
+        let tx_dst = T::regs().txdr().ptr() as *mut u8;
+        let clock_byte = 0x00;
+        let tx_f = self
+            .txdma
+            .write_x(tx_request, &clock_byte, clock_byte_count, tx_dst);
+
+        unsafe {
+            T::regs().cfg1().modify(|reg| {
+                reg.set_txdmaen(true);
+            });
+            T::regs().cr1().modify(|w| {
+                w.set_spe(true);
+            });
+            T::regs().cr1().modify(|w| {
+                w.set_cstart(true);
+            });
+        }
+
+        join3(tx_f, rx_f, Self::wait_for_idle()).await;
+        unsafe {
+            T::regs().cfg1().modify(|reg| {
+                reg.set_rxdmaen(false);
+                reg.set_txdmaen(false);
+            });
+            T::regs().cr1().modify(|w| {
+                w.set_spe(false);
+            });
+        }
+        Ok(())
+    }
+
+    #[allow(unused)]
+    async fn read_write_dma_u8(&mut self, read: &mut [u8], write: &[u8]) -> Result<(), Error>
+    where
+        Tx: TxDmaChannel<T>,
+        Rx: RxDmaChannel<T>,
+    {
+        Self::set_word_size(WordSize::EightBit);
+        unsafe {
+            T::regs().cr1().modify(|w| {
+                w.set_spe(false);
+            });
+            T::regs().cfg1().modify(|reg| {
+                reg.set_rxdmaen(true);
+            });
+        }
+
+        let rx_request = self.rxdma.request();
+        let rx_src = T::regs().rxdr().ptr() as *mut u8;
+        let rx_f = self
+            .rxdma
+            .read(rx_request, rx_src, &mut read[0..write.len()]);
```

Comment:
This seems to make more sense if compared to the other two implementations.

Reply:
Maybe we should have some indications that says that if `read.len() > write.len()` then `read` won't be fully filled ? Or maybe just require the lengths to be equal ?

Reply:
Actually, I think that might panic if `write` is bigger than `read`.

Reply:
Yah I think definitely an assert but I am not sure how I feel about the slicing really. 

I'll double check which DMA directions get enabled in write-only modes on all of them.

Reply:
If read is longer then it never fills and won't complete. Which I hit a few times myself.